tasks_week1_Gal_Malka/src/Vector.cpp

```diff
@@ -0,0 +1,355 @@
+#include "Vector.h"
+
+/*
+* Main consturctor - gets size and value and initalizes the vector.
+*/
+Vector::Vector(size_t size = 0, int value = 0) : _size(size), _capacity(size) {
+
+	_data = new int[_capacity];
+	for (size_t index = 0; index < _size; index++) {
+		_data[index] = value;
+	}
+
+}
+
+/*
+* Size consturctor - gets size and initalizes the vector with size * elements with the value 0.
+*/
+Vector::Vector(size_t size) : _size(size), _capacity(size) {
+
+	_data = new int[_capacity];
+
+	for (size_t index = 0; index < _size; index++) {
+		_data[index] = 0;
+	}
+}
+
+/*
+* Copy consturctor - copy the argument vector data into the created vector.
+*/
+Vector::Vector(const Vector& other) {
+
+	_size = other._size;
+	_capacity = other._capacity;
+	_data = new int[_capacity];
+	copy_data(other._data, &_data, other._size);
+
+}
+
+/*
+* Move consturctor - move the argument vector's data control into the created vector.
+*/
+Vector::Vector(Vector&& other) noexcept {
+
+	_size = other._size;
+	_capacity = other._capacity;
+	_data = other._data;
+	other._data = nullptr;
+}
+
+/*
+* Initializer list consturctor - construct the vector's data from initializer values.
+*/
+Vector::Vector(const std::initializer_list<int> initializer) {
+
+	_size = initializer.size();
+	_capacity = _size;
+	_data = new int[_capacity];
+	copy_data(initializer.begin(), &_data, _size);
+}
+
+/*
+* Destructor - deallocate vector's memory
+*/
+Vector::~Vector() {
+
+	if (_data) {
+
+		delete[] _data;
+	}
+}
+
+/*
+* operator[] - for random access to vector's values.
+*/
+int& Vector::operator[] (size_t index) {
+
+	if (index < 0 || index > _size - 1) {
+
+		throw std::out_of_range("Array index out of bounds.");
+	}
+	return *(_data + index);
+}
+
+/*
+* const operator[] - for random access to vector's values without changing any of the data.
+*/
+const int& Vector::operator[] (size_t index) const {
+
+	if (index < 0 || index > _size - 1) {
+
+		throw std::out_of_range("Array index out of bounds.");
+	}
+
+	return *(_data + index);
+}
+
+/*
+* operator= - copy the argument vector data into the vector and deallocate previous data.
+*/
+Vector& Vector::operator= (const Vector& other) {
+
+	if (this != &other) {
+
+		_size = other._size;

```

Comment:
There is a more elegant way to implement this. 
Check out the "copy-and-swap idiom" 
https://stackoverflow.com/questions/3279543/what-is-the-copy-and-swap-idiom